### PR TITLE
Feat/6487 part2 silver transformations

### PR DESCRIPTION
**Pull request by motylp**

### CODE REVIEW

File: pysparkle/pysparkle/etl.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class TableTransformation:
+    table_name: str
+    transformation_function: Callable[[DataFrame], DataFrame]
+
+
+def save_files_as_delta_tables(
```
Comment (motylp) [inline pysparkle/pysparkle/etl.py:27]: no change in this function - it was just moved from etl.etl.py to etl.py

**Review by lorraine-houston-ensono — APPROVED** [review]

File: de_workloads/data_processing/silver/spark_jobs/silver.py
[the remark this answers is not in the document]
Comment (adurkan-amido) [reply de_workloads/data_processing/silver/spark_jobs/silver.py:1]: If we do, will also need to update the DevOps pipeline - happy for this rename to be picked up in another ticket if we think it's worthwhile.

File: docs/workloads/common/data/pysparkle/pysparkle_quickstart.md [hunk 2]
```diff
@@ -21,87 +21,32 @@ data processing workflows with Apache Spark, and is an integral part of the Stac
...
-various formats (e.g. CSV, Parquet, JSON, Delta) as an input and converts it into a more structured Delta format.
-- **Gold processing**: Capabilities for performing 'Silver to Gold' layer transformations. Supports additional
-transformations and aggregations of data, creating data sets optimised for downstream consumption.
-- **Data Quality**: PySparkle performs data quality checks according to user-provided configurations, ensuring the
-reliability and accuracy of the data.
+- **ETL Transformation Utilities** which streamline Data Pipelines by:
+    - Offering PySpark utility functions. They accept raw data in various formats (e.g., CSV, Parquet, JSON, Delta)
+    and convert it into a structured Delta format. PySparkle also facilitates pre-defined or user-defined transformations
+    on dataframes.
+    - Providing storage utilities that ensure a smooth interaction with ADLS and Blob storage.
```
Comment (adurkan-amido) [inline docs/workloads/common/data/pysparkle/pysparkle_quickstart.md:28]:
```suggestion
    - Providing storage utilities that ensure smooth and secure interaction with Azure Data Lake and Blob storage.
```

File: pysparkle/README.md [hunk 3]
```diff
@@ -8,87 +8,32 @@ data processing workflows with Apache Spark, and is an integral part of the Stac
...
-various formats (e.g. CSV, Parquet, JSON, Delta) as an input and converts it into a more structured Delta format.
-- **Gold processing**: Capabilities for performing 'Silver to Gold' layer transformations. Supports additional
-transformations and aggregations of data, creating data sets optimised for downstream consumption.
-- **Data Quality**: PySparkle performs data quality checks according to user-provided configurations, ensuring the
-reliability and accuracy of the data.
+- **ETL Transformation Utilities** which streamline Data Pipelines by:
+    - Offering PySpark utility functions. They accept raw data in various formats (e.g., CSV, Parquet, JSON, Delta)
+    and convert it into a structured Delta format. PySparkle also facilitates pre-defined or user-defined transformations
+    on dataframes.
+    - Providing storage utilities that ensure a smooth interaction with ADLS and Blob storage.
```
Comment (adurkan-amido) [inline pysparkle/README.md:15]: As above

File: pysparkle/README.md [hunk 4]
```diff
@@ -8,87 +8,32 @@ data processing workflows with Apache Spark, and is an integral part of the Stac
...
 The key components of PySparkle include:
 
-- **Silver processing**: Capabilities for performing 'Bronze to Silver' layer transformations. It accepts raw data of
-various formats (e.g. CSV, Parquet, JSON, Delta) as an input and converts it into a more structured Delta format.
-- **Gold processing**: Capabilities for performing 'Silver to Gold' layer transformations. Supports additional
-transformations and aggregations of data, creating data sets optimised for downstream consumption.
-- **Data Quality**: PySparkle performs data quality checks according to user-provided configurations, ensuring the
-reliability and accuracy of the data.
+- **ETL Transformation Utilities** which streamline Data Pipelines by:
+    - Offering PySpark utility functions. They accept raw data in various formats (e.g., CSV, Parquet, JSON, Delta)
```
Comment (adurkan-amido) [inline pysparkle/README.md:12]:
```suggestion
    - Offering PySpark utility functions. They accept raw data in various formats (e.g. CSV, Parquet, JSON, Delta)
```

File: pysparkle/pysparkle/etl.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        transform_func: Transformation function.
+        target_container: Name of the destination container in ADLS.
+        output_file_name: The name of the file (including any subdirectories within the container).
+        overwrite: Flag to determine whether to overwrite the entire table or perform an upsert.
+        merge_keys: List of keys based on which upsert will be performed.
+
+    """
+    transformed_df = transform_func(input_df)
+    output_filepath = get_adls_file_url(target_container, output_file_name)
+    save_dataframe_as_delta(spark, transformed_df, output_filepath, overwrite, merge_keys)
```
Comment (adurkan-amido) [inline pysparkle/pysparkle/etl.py:141]: We should add some metadata columns to the output table, similar to the ones in the ingest table.
(Happy for this to be a separate task / PR)

File: de_workloads/data_processing/silver/spark_jobs/silver.py
[the remark this answers is not in the document]
Comment (motylp) [reply de_workloads/data_processing/silver/spark_jobs/silver.py:1]: I'm fine with renaming this in a separate task: https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6723. When doing this we might also think about unifying naming conventions (e.g. Ingest_AzureSql_Example vs silver_movies_example).

File: de_workloads/data_processing/silver/spark_jobs/silver.py [hunk 6]
```diff
@@ -1,42 +1,144 @@
...
+    )
+
+    return rename_columns_to_snake_case(df_final)
+
+
+def etl_main() -> None:
+    """Execute the Silver processing for a given dataset."""
     logger.info("Running Silver processing...")
 
     spark = get_spark_session_for_adls("Silver")
```
[the remark this answers is not in the document]
Comment (motylp) [reply de_workloads/data_processing/silver/spark_jobs/silver.py]: done

File: de_workloads/data_processing/silver/spark_jobs/silver.py [hunk 7]
```diff
@@ -1,42 +1,144 @@
...
+        .withColumn("production_country_iso", col("production_country.iso_3166_1"))
+        .withColumn("spoken_language_iso", col("spoken_language.iso_639_1"))
+        .drop("genre", "production_company", "production_country", "spoken_language")
+    )
+
+    return rename_columns_to_snake_case(df_final)
+
+
+def etl_main() -> None:
+    """Execute the Silver processing for a given dataset."""
```
[the remark this answers is not in the document]
Comment (motylp) [reply de_workloads/data_processing/silver/spark_jobs/silver.py]: done

File: de_workloads/data_processing/silver/spark_jobs/silver.py [hunk 8]
```diff
@@ -1,42 +1,144 @@
...
+            BRONZE_CONTAINER,
+            INPUT_PATH_PATTERN.format(table_name=table.table_name),
+            datasource_type=SOURCE_DATA_TYPE,
+        )
+
+        output_path = OUTPUT_PATH_PATTERN.format(table_name=table.table_name)
+
+        transform_and_save_as_delta(spark, df, table.transformation_function, SILVER_CONTAINER, output_path)
 
     logger.info("Finished: Silver processing.")
```
[the remark this answers is not in the document]
Comment (motylp) [reply de_workloads/data_processing/silver/spark_jobs/silver.py]: done

File: docs/workloads/azure/data/etl_pipelines/silver_data_azure.md [hunk 9]
```diff
@@ -40,10 +40,12 @@ The diagram below gives an overview of the Silver pipeline design.
 
 ![ADF_SilverPipelineDesign.png](../images/ADF_SilverPipelineDesign.png)
 
-
 Silver processing is executed as Python Databricks job, with repeatable data transformation processes packaged within
 our [PySparkle](../../../common/data/pysparkle/pysparkle_quickstart.md) library.
 
+The logic specific to a particular pipeline is kept inside a `spark_jobs` directory together with all the other pipeline
+definitions. For details please review the contents of [silver.py](https://github.com/amido/stacks-azure-data/blob/main/de_workloads/data_processing/silver/spark_jobs/silver.py).
```
[the remark this answers is not in the document]
Comment (motylp) [reply docs/workloads/azure/data/etl_pipelines/silver_data_azure.md:47]: Task created: https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6723

File: docs/workloads/azure/data/etl_pipelines/silver_data_azure.md [hunk 10]
```diff
@@ -58,18 +60,6 @@ In ADF the base Silver pipeline is as simple as this:
...
-from pysparkle.pysparkle_cli import cli
-
-def call_pysparkle_entrypoint():
-    cli(["silver", "--dataset-name", "movies_dataset"], standalone_mode=False)
-
-if __name__ == "__main__":
-    call_pysparkle_entrypoint()
-```
+It contains just one step - Python Databricks, configured to run a `silver.py` script, which gets deployed to DBFS
+(`dbfs:/FileStore/scripts/silver/silver.py`). Datastacks library with PySparkle package is attached to a cluster by
```
[the remark this answers is not in the document]
Comment (motylp) [reply docs/workloads/azure/data/etl_pipelines/silver_data_azure.md:64]: same as above

File: docs/workloads/common/data/datastacks.md [hunk 11]
```diff
@@ -100,3 +98,17 @@ default_arm_deployment_mode: Incremental
 window_start_default: 2010-01-01
 window_end_default: 2010-01-31
 ```
+
+## Performing data quality checks
+
+Datastacks provides a CLI to conduct data quality checks using [PySparkle](pysparkle/pysparkle_quickstart.md) library as a backend.
```
[the remark this answers is not in the document]
Comment (motylp) [reply docs/workloads/common/data/datastacks.md]: done

File: docs/workloads/common/data/pysparkle/pysparkle_quickstart.md (hunk 2 above)
Comment (motylp) [reply docs/workloads/common/data/pysparkle/pysparkle_quickstart.md]: done

File: pysparkle/README.md (hunk 3 above)
Comment (motylp) [reply pysparkle/README.md]: done

File: pysparkle/README.md (hunk 4 above)
Comment (motylp) [reply pysparkle/README.md]: done

File: pysparkle/pysparkle/pyspark_utils.py [hunk 12]
```diff
@@ -117,3 +151,17 @@ def ensure_database_exists(spark: SparkSession, schema: str) -> None:
     if not spark.catalog.databaseExists(schema):
         spark.sql(f"CREATE DATABASE IF NOT EXISTS {schema}")
         logger.info(f"Database {schema} has been created.")
+
+
+def rename_columns_to_snake_case(df: DataFrame) -> DataFrame:
+    """Rename all columns of a DataFrame to snake_case.
```
[the remark this answers is not in the document]
Comment (motylp) [reply pysparkle/pysparkle/pyspark_utils.py]: done

File: de_templates/test_config_ingest.yaml [hunk 13]
```diff
@@ -23,11 +23,13 @@ ado_variable_groups_prod:
 
 
 # Datalake containers
-
 bronze_container: raw
 silver_container: staging
 gold_container: curated
 
+# Stores configurations in Azure Blob storage
+config_container: config
```
[the remark this answers is not in the document]
Comment (motylp) [reply de_templates/test_config_ingest.yaml:31]: Yeah, that is a good remark. During the deployment the config container is as customisable as the ADLS containers. None of our containers is 100% customisable:
- In shared resources we have hardcoded "raw" container
- "config" is additionally hardcoded in the pipelines, so the question is do we want it to be customisable?

I will revert config customisation for now, as making it configurable would require more changes.

File: pysparkle/pysparkle/etl.py (hunk 5 above)
Comment (motylp) [reply pysparkle/pysparkle/etl.py:141]: Task created: https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6724

File: de_templates/test_config_ingest.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (motylp) [reply de_templates/test_config_ingest.yaml]: Done

File: de_workloads/data_processing/silver/spark_jobs/silver.py [hunk 14]
```diff
@@ -1,42 +1,144 @@
...
+    read_latest_rundate_data,
+    transform_and_save_as_delta,
+)
+from pysparkle.logger import setup_logger
+from pysparkle.pyspark_utils import rename_columns_to_snake_case
 
 BRONZE_CONTAINER = "raw"
 SILVER_CONTAINER = "staging"
+SOURCE_DATA_TYPE = "parquet"
+INPUT_PATH_PATTERN = "Ingest_AzureSql_Example/movies.{table_name}/v1/full/"
```
[the remark this answers is not in the document]
Comment (motylp) [reply de_workloads/data_processing/silver/spark_jobs/silver.py:20]: User story created for this: https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6728

File: de_templates/test_config_ingest.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (adurkan-amido) [reply de_templates/test_config_ingest.yaml]: I think leaving it as 'config' is fine.

File: docs/workloads/azure/data/etl_pipelines/silver_data_azure.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (adurkan-amido) [reply docs/workloads/azure/data/etl_pipelines/silver_data_azure.md:47]: Were you happy with the wording change suggestion?

File: pysparkle/pysparkle/etl.py (hunk 5 above)
Comment (adurkan-amido) [reply pysparkle/pysparkle/etl.py:141]: Added here - https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6724

File: docs/workloads/azure/data/etl_pipelines/silver_data_azure.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (adurkan-amido) [reply docs/workloads/azure/data/etl_pipelines/silver_data_azure.md:47]: I mean the one in this comment, as it's not been committed 🙂

[the remark this answers is not in the document]
Comment (motylp) [reply docs/workloads/azure/data/etl_pipelines/silver_data_azure.md]: I was a bit in a hurry and missed that :D should be good now!

**Review by adurkan-amido — APPROVED** [review]

**Pull request merged**